apps/prairielearn/src/question-servers/freeform.ts

```diff
@@ -2080,20 +2084,46 @@ async function getContext(question: Question, course: Course): Promise<QuestionP
 
 async function getCacheKey(
   course: Course,
+  variant: Variant,
+  submission: Submission | null,
   data: ExecutionData,
   context: QuestionProcessingContext,
-) {
+): Promise<string | null> {
   try {
     const commitHash = await getOrUpdateCourseCommitHash(course);
-    const dataHash = objectHash({ data, context }, { algorithm: 'sha1', encoding: 'base64' });
-    return `question:${commitHash}-${dataHash}`;
+
+    const dataHash = objectHash(
+      [
+        // We deliberately exclude large user-controlled objects from the cache key.
+        // Whenever these change, the `modified_at` column of `variants` and/or
+        // `submissions` will change, which will cause the cache to be invalidated.
+        _.omit(data, [
+          'params',
+          'correct_answers',
+          'submitted_answers',
+          'format_errors',
+          'partial_scores',
+          'raw_submitted_answers',
+        ]),
+        context,
+        variant.modified_at,
+        submission?.modified_at,
+      ],
+      { algorithm: 'sha1', encoding: 'base64' },
+    );
+
+    // The variant and submission IDs are included in the cache key to ensure
+    // that data never leaks between variants or submissions.
+    return `variant:${variant.id}:submission:${submission?.id}:${commitHash}-${dataHash}:cache`;
```

Comment:
```suggestion
    return `variant:${variant.id}:submission:${submission?.id ?? null}:${commitHash}-${dataHash}:cache`;
```

Reply:
The other comment says 

> for the price of a few extra bytes in the key, we can be really sure that we're not accidentally getting cache hits when we shouldn't.

In what scenarios would this be null? I wouldn't want an accidental cache collision between two submissions with `null` IDs.

Reply:
`variants.id` is the primary key of the table, so yep, we don't have to worry about collisions there.